In [103]:
# Установить необходимый пакет для работы с URL
install.packages("readr")

# Загрузить библиотеку
library(readr)

In [104]:
# Указать URL-адрес файла CSV на GitHub
url <- "https://raw.githubusercontent.com/kindykbayev/KIS/main/lab1/Computers.csv"

In [105]:
Sys.setenv("VROOM_CONNECTION_SIZE" = 1048576)

In [106]:
# Прочитать CSV-файл
data <- read_csv(url, col_types = cols(.default = "integer", cd = col_character(), multi = col_character(), premium = col_character()))

New names:
• `` -> `...1`


In [107]:
View(data)

...1 price speed hd   ram screen cd  multi premium ads trend
1     1   1499  25     80   4  14     no  no    yes     94  1    
2     2   1795  33     85   2  14     no  no    yes     94  1    
3     3   1595  25    170   4  15     no  no    yes     94  1    
4     4   1849  25    170   8  14     no  no    no      94  1    
5     5   3295  33    340  16  14     no  no    yes     94  1    
6     6   3695  66    340  16  14     no  no    yes     94  1    
7     7   1720  25    170   4  14     yes no    yes     94  1    
8     8   1995  50     85   2  14     no  no    yes     94  1    
9     9   2225  50    210   8  14     no  no    yes     94  1    
10   10   2575  50    210   4  15     no  no    yes     94  1    
11   11   2195  33    170   8  15     no  no    yes     94  1    
12   12   2605  66    210   8  14     no  no    yes     94  1    
13   13   2045  50    130   4  14     no  no    yes     94  1    
14   14   2295  25    245   8  14     no  no    yes     94  1    
15   15   2699  50    212   8  14     no  no    yes     94  1    
16   16   2225  50    130   4  14     no  no    yes     94  1    
17   17   1595  33     85   2  14     no  no    yes     94  1    
18   18   2325  33    210   4  15     no  no    yes     94  1    
19   19   2095  33    250   4  15     no  no    yes     94  1    
20   20   4395  66    452   8  14     no  no    yes     94  1    
21   21   1695  33    130   4  14     no  no    yes     94  1    
22   22   2795  66    130   4  14     no  no    yes     94  1    
23   23   2895  25    340  16  14     no  no    yes     94  1    
24   24   2875  50    210   4  17     no  no    yes     94  1    
25   25   4195  50    452   8  14     no  no    yes     94  1    
26   26   1290  33     80   2  14     no  no    yes     94  1    
27   27   1975  33    130   4  14     no  no    yes     94  1    
28   28   3995  33    452   8  14     no  no    yes     94  1    
29   29   3095  33    340  16  14     no  no    yes     94  1    
30   30   3244  66    245   8  14     no  no    yes     94  1    
⋮    ⋮    ⋮     ⋮     ⋮    ⋮   ⋮      ⋮   ⋮     ⋮       ⋮   ⋮    
6230 6230 2344   66    850 16  15     yes no    yes     39  35   
6231 6231 2890  100   1200 24  17     yes no    yes     39  35   
6232 6232 2390   66    850 16  17     yes no    yes     39  35   
6233 6233 1495   66    528  8  14     no  no    yes     39  35   
6234 6234 1844  100    528  8  14     no  no    yes     39  35   
6235 6235 1699   66    630  8  15     no  no    no      39  35   
6236 6236 3290   75   1600 32  17     yes no    yes     39  35   
6237 6237 2790   66   1200 24  17     yes no    yes     39  35   
6238 6238 2845  100   1600 24  15     yes no    yes     39  35   
6239 6239 1744   66    528  8  14     no  no    yes     39  35   
6240 6240 3095  100   1600 32  15     yes no    yes     39  35   
6241 6241 1595  100    528  8  14     no  no    yes     39  35   
6242 6242 2254  100    850 16  15     yes no    yes     39  35   
6243 6243 2723  100   1200 24  15     yes yes   yes     39  35   
6244 6244 2745   66   1600 24  15     yes no    yes     39  35   
6245 6245 3054   75   1600 32  15     yes no    yes     39  35   
6246 6246 2744   66   1200 24  15     yes no    yes     39  35   
6247 6247 1554   66    528  8  14     no  no    yes     39  35   
6248 6248 1654  100    528  8  14     no  no    yes     39  35   
6249 6249 2623   66   1200 24  15     yes yes   yes     39  35   
6250 6250 2095   66    850 16  15     yes no    yes     39  35   
6251 6251 2495   66   1200 24  15     yes no    yes     39  35   
6252 6252 2995   66   1600 32  15     yes no    yes     39  35   
6253 6253 1690   66   1200  8  14     no  no    yes     39  35   
6254 6254 2154   66    850 16  15     yes no    yes     39  35   
6255 6255 1690  100    528  8  15     no  no    yes     39  35   
6256 6256 2223   66    850 16  15     yes yes   yes     39  35   
6257 6257 2654  100   1200 24  15     yes no    yes     39  35   
6258 6258 2195  100    850 16  15     yes no 

In [108]:
set.seed(13)
Count_low<-data.frame()
for (i in c(1:50)){
    Count_low<-rbind(Count_low, c(i,sample(c(3:6),1), sample(c(6:20),1), sample(c(1:20),1), sample(c(1:9),1)))}
colnames(Count_low)<-c("Вариант","Количество С1 >=","Количество С2 >=", "Количество С3 >=", "Количество С4 >=")
View(Count_low[6, ])

Вариант Количество С1 >= Количество С2 >= Количество С3 >= Количество С4 >=
6 6       6                6                11               4

In [109]:
Count_up<-data.frame()
for (i in c(1:50)){
    Count_up<-rbind(Count_up, c(i,Count_low[i,2]+sample(c(3:6),1), Count_low[i,3]+sample(c(6:20),1),
                                          Count_low[i,4]+sample(c(1:20),1), Count_low[i,5]+sample(c(1:9),1)))}
colnames(Count_up)<-c("Вариант","Количество С1 <=","Количество С2 <=", "Количество С3 <=", "Количество С4 <=")
View(Count_up[6, ])

Вариант Количество С1 <= Количество С2 <= Количество С3 <= Количество С4 <=
6 6       10               22               15               13

In [110]:
Res<-lm(price~speed+ram+hd, data)
Res


Call:
lm(formula = price ~ speed + ram + hd, data = data)

Coefficients:
(Intercept)        speed          ram           hd  
  1492.1775       5.8112      78.2357      -0.5357  


In [111]:
g<-predict(Res,data)
View(data.frame(data$price, g))

data.price g       
1    1499       1907.542
2    1795       1794.882
3    1595       1859.328
4    1849       2172.271
5    3295       2753.573
6    3695       2945.343
7    1720       1859.328
8    1995       1893.672
9    2225       2296.121
10   2575       1983.179
11   2195       2218.760
12   2605       2389.100
13   2045       2026.036
14   2295       2132.092
15   2699       2295.050
16   2225       2026.036
17   1595       1794.882
18   2325       1884.388
19   2095       1862.960
20   4395       2259.456
21   1695       1927.246
22   2795       2119.015
23   2895       2707.084
24   2875       1983.179
25   4195       2166.477
26   1290       1797.561
27   1975       1927.246
28   3995       2067.687
29   3095       2753.573
30   3244       2370.350
⋮    ⋮          ⋮       
6230 2344       2672.125
6231 2890       3308.089
6232 2390       2672.125
6233 1495       2218.741
6234 1844       2416.322
6235 1699       2164.098
6236 3290       3574.407
6237 2790       3110.509
6238 2845       3093.801
6239 1744       2218.741
6240 3095       3719.687
6241 1595       2416.322
6242 2254       2869.706
6243 2723       3308.089
6244 2745       2896.221
6245 3054       3574.407
6246 2744       3110.509
6247 1554       2218.741
6248 1654       2416.322
6249 2623       3110.509
6250 2095       2672.125
6251 2495       3110.509
6252 2995       3522.106
6253 1690       1858.737
6254 2154       2672.125
6255 1690       2416.322
6256 2223       2672.125
6257 2654       3308.089
6258 2195       2869.706
6259 2490       2869.706

Существует 4 сегмента в которые требуется выбрать кол-во компьютеров установленное вариантом:
$$6<=С1<=10$$
$$6<=С2<=22$$
$$11<=С3<=15$$
$$4<=С4<=13$$

Стоимость отдельного компьютера от конструктивных параметров  𝑠𝑝𝑒𝑒𝑑,𝑟𝑎𝑚,ℎ𝑑  выражена ранее через регрессионную зависимость:
$$𝑝𝑟𝑖𝑐𝑒=1492.1775+5.8112⋅𝑠𝑝𝑒𝑒𝑑+78.2357⋅𝑟𝑎𝑚−0.5357⋅ℎ𝑑$$

Добавим ограничения нижнего уровня:
 - суммарная стоимость компьютеров сегмента $C_1$ должна быть не более 15000 у.е.
 - суммарная производительность $C_2$ не менее 300 (считаем аддитивной характеристикой от количества входящих ПК)
 - максимальный объем ОЗУ для сегмента $C_3$ не более 8
 - минимальный объем жесткого диска для сегмента для сегмента $C_4$ не менее 1000
 - Суммарная производительность для С4 - не менее 700 (считаем аддитивной характеристикой от количества входящих ПК)

Введём переменные:

$N_1$ - количество серверов

$N_2$ - количество компьютеров сегмента аналитиков и менеджеров

$N_3$ - количество компьютеров сегмента обслуживающего персонала

$N_4$ - количество компьютеров сегмента с видео, аудио оборудованием

$price$ - стоимость компьютера без учёта характеристик

$speed$ - зависимость стоимости компьютера от его скорости

$ram$ - зависимость стоимости компьютера от его оперативной памяти

$hd$ - зависимость стоимости компьютера от объёма диска

Критерий верхнего уровня:

$$𝐻1=𝑁1⋅𝑝𝑟𝑖𝑐𝑒1+𝑁2⋅𝑝𝑟𝑖𝑐𝑒2+𝑁3⋅𝑝𝑟𝑖𝑐𝑒3+𝑁4⋅𝑝𝑟𝑖𝑐𝑒4→min$$

После подстановки регрессионных зависимостей получаем:

$$𝐻_1=1492.1775𝑁_1+5.8112⋅𝑠𝑝𝑒𝑒𝑑_1⋅𝑁_1+78.2357⋅𝑟𝑎𝑚_1⋅𝑁_1−0.5357⋅ℎ𝑑_1⋅𝑁_1$$

$$+1492.1775𝑁_2+5.8112⋅𝑠𝑝𝑒𝑒𝑑_2⋅𝑁_2+78.2357⋅𝑟𝑎𝑚_2⋅𝑁_2−0.5357⋅ℎ𝑑_2⋅𝑁_2$$

$$+1492.1775𝑁_3+5.8112⋅𝑠𝑝𝑒𝑒𝑑_3⋅𝑁_3+78.2357⋅𝑟𝑎𝑚_3⋅𝑁_3−0.5357⋅ℎ𝑑_3⋅𝑁_3$$

$$+1492.1775𝑁_4+5.8112⋅𝑠𝑝𝑒𝑒𝑑_4⋅𝑁_4
+78.2357⋅𝑟𝑎𝑚_4⋅𝑁_4−0.5357⋅ℎ𝑑_4⋅𝑁_4$$

Обозначим:
$$1492.1775N_1 = p_1,$$
$$6\cdot1492.1775\leq p_1\leq 10\cdot 1492.1775$$
$$speed_1\cdot N_1 = s_1$$
$$ram_1\cdot N_1 = r_1$$
$$ hd_1\cdot N_1 = h_1$$
___
$$1492.1775N_2 = p_2,$$
$$6\cdot 1492.1775\leq p_2\leq 22\cdot 1492.1775$$
$$speed_2\cdot N_2 = s_2$$
$$ram_2\cdot N_2 = r_2$$
$$hd_2\cdot N_2 = h_2$$
___
$$1492.1775N_3 = p_3,$$
$$11\cdot 1492.1775\leq p_3\leq 15\cdot 1492.1775$$
$$speed_3\cdot N_3 = s_3$$
$$ram_3\cdot N_3 = r_3$$
$$hd_3\cdot N_3 = h_3$$
___
$$1492.1775N_4 = p_4,$$
$$4\cdot 1492.1775\leq p_4\leq 13\cdot 1492.1775$$
$$speed_2\cdot N_4 = s_4$$
$$ram_2\cdot N_4 = r_4$$
$$hd_2\cdot N_4 = h_4$$

Для нахождения ограничений на остальные  $𝑥_i$  можно взять усредненные характеристики:

In [112]:
mean(data$speed)
mean(data$ram)
mean(data$hd)

[1] 52.01102

[1] 8.286947

[1] 416.6017

Соответственно остальные ограничения следующие:
<center>price</center>
$$6\cdot 1492.1775\leq p_{1}\leq 10 \cdot 1492.1775$$
<center>speed</center>
$$6\cdot 52.011\leq s_1\leq 10\cdot 52.011$$
<center>ram</center>
$$6\cdot 8.2869\leq r_1\leq 10\cdot 8.2869$$
<center>hd</center>
$$6\cdot 416.6017\leq h_1\leq 10\cdot 416.6017$$


---



<center>price</center>
$$6\cdot 1492.1775\leq p_{2}\leq 22\cdot 1492.1775$$
<center>speed</center>
$$6\cdot 52.011\leq s_2\leq 22\cdot 52.011$$
<center>ram</center>
$$6\cdot 8.2869\leq r_2\leq 22\cdot 8.2869$$
<center>hd</center>
$$6\cdot 416.6017\leq h_2\leq 22\cdot 416.6017$$


---



<center>price</center>
$$11\cdot 1492.1775\leq p_{3}\leq 15\cdot 1492.1775$$
<center>speed</center>
$$11\cdot 52.011\leq s_3\leq 15\cdot 52.011$$
<center>ram</center>
$$11\cdot 8.2869\leq r_3\leq 15\cdot 8.2869$$
<center>hd</center>
$$11\cdot 416.6017\leq h_3\leq 15\cdot 416.6017$$


---



<center>price</center>
$$4\cdot 1492.1775\leq p_{4}\leq 13\cdot 1492.1775$$
<center>speed</center>
$$4\cdot 52.011\leq s_4\leq 13\cdot 52.011$$
<center>ram</center>
$$4\cdot 8.2869\leq r_4\leq 13\cdot 8.2869$$
<center>hd</center>
$$4\cdot 416.6017\leq h_4\leq 13\cdot 416.6017$$


Учтем ограничения нижнего уровня:

$$p_{1} + 5.8112\cdot s_{1} + 78.2357\cdot r_{1} - 0.5357\cdot h_{1} \leq 10000$$
$$s_{2} + r_{2} \geq 300$$
$${𝑟𝑎𝑚}_3 \leq 8$$
$${hd}_4 \geq 1000$$
$$s_{4} + r_{4} \geq 700$$

Проведём нормализацию $s_2$ и $r_2$ для $С_2$ и $s_4$ и $r_4$ для $С_4$, для этого необходимо найти максимальные значения speed, ram, $N_2$ и $N_4$.

In [113]:
max(data$speed)
max(data$ram)

[1] 100

[1] 32

$$\bar{s}_{2}=\frac{s_{2}}{max(N_2)\cdot max(speed)}=\frac{s_{2}}{22\cdot 100}=\frac{s_{2}}{2200}$$
$$\bar{r}_{2}=\frac{r_{2}}{max(N_2)\cdot max(ram)}=\frac{r_{2}}{22\cdot 32}=\frac{r_{2}}{704}$$

$$\bar{s}_{4}=\frac{s_{4}}{max(N_4)\cdot max(speed)}=\frac{s_{4}}{13\cdot 100}=\frac{s_{4}}{1300}$$
$$\bar{r}_{4}=\frac{r_{4}}{max(N_4)\cdot max(ram)}=\frac{r_{4}}{13\cdot 32}=\frac{r_{4}}{416}$$

Сформулируем также частные критерии оптимизации на нижнем уровне:
Cуммарная производительность (speed, ram) в сегменте  $C_2$ и $C_4$  максимизируется.
#Для сегмента $C_2$:
$$W_2 = \alpha\cdot s_2+\beta \cdot r_2\rightarrow \max$$
Подставим нормализованные значения:
$$W_2 = \alpha\cdot \frac{s_{2}}{2200}+\beta \cdot \frac{r_{2}}{704}\rightarrow \max$$
Предположим что объем оперативной памяти и скорость имеют одинаковую важность для рабочих станций аналитиков и менеджеров, поэтому: $\alpha = 0.5, \beta = 0.5$:
$$W_2 = 0.5\cdot \frac{s_{2}}{2200} + 0.5 \cdot \frac{r_{2}}{704}\rightarrow \max$$
#Для сегмента $C_4$:
$$W_4 = \alpha\cdot s_4+\beta \cdot r_4\rightarrow \max$$
Подставим нормализованные значения:
$$W_4 = \alpha\cdot \frac{s_{4}}{1300}+\beta \cdot \frac{r_{4}}{416}\rightarrow \max$$
Предположим для станций сопряжения с видео и аудио оборудованием скорость имееет больший приоритет: $\alpha = 0.6, \beta = 0.4$:
$$W_4 = 0.6\cdot \frac{s_{4}}{1300} + 0.4 \cdot \frac{r_{4}}{416}\rightarrow \max$$
#Формируем критерий нижнего уровня:
$$H_2=0.5 W_1+0.5\cdot W_2= 0.5\cdot 0.5\cdot \frac{s_{2}}{2200} + 0.5\cdot 0.5 \cdot \frac{r_{2}}{704} + W_4 = 0.5\cdot 0.6\cdot \frac{s_{4}}{1300} + 0.5\cdot 0.4 \cdot \frac{r_{4}}{416}\rightarrow \max$$

#Окончательно получаем:
$$𝐻_1={p_{1}}+5.8112⋅{s_{1}}+78.2357{r_{1}}−0.5357⋅{h_{1}}$$

$$+{p_{2}}+5.8112⋅{s_{2}}+78.2357⋅{r_{2}}−0.5357⋅{h_{2}}$$

$$+{p_{3}}+5.8112⋅{s_{3}}+78.2357⋅{r_{3}}−0.5357⋅{h_{3}}$$

$$+{p_{4}}+5.8112⋅{s_{4}}+78.2357⋅{r_{4}}−0.5357⋅{h_{4}}$$
___
$$H_2= 0.5\cdot 0.5\cdot \frac{s_{2}}{2200} + 0.5\cdot 0.5 \cdot \frac{r_{2}}{704} + 0.5\cdot 0.6\cdot \frac{s_{4}}{1300} + 0.5\cdot 0.4 \cdot \frac{r_{4}}{416}\rightarrow \max$$
___
<center>price</center>
$$6\cdot 1492.1775\leq p_{1}\leq 10 \cdot 1492.1775$$
<center>speed</center>
$$6\cdot 52.011\leq s_1\leq 10\cdot 52.011$$
<center>ram</center>
$$6\cdot 8.2869\leq r_1\leq 10\cdot 8.2869$$
<center>hd</center>
$$6\cdot 416.6017\leq h_1\leq 10\cdot 416.6017$$

---

<center>price</center>
$$6\cdot 1492.1775\leq p_{2}\leq 22\cdot 1492.1775$$
<center>speed</center>
$$6\cdot 52.011\leq s_2\leq 22\cdot 52.011$$
<center>ram</center>
$$6\cdot 8.2869\leq r_2\leq 22\cdot 8.2869$$
<center>hd</center>
$$6\cdot 416.6017\leq h_2\leq 22\cdot 416.6017$$

---

<center>price</center>
$$11\cdot 1492.1775\leq p_{3}\leq 15\cdot 1492.1775$$
<center>speed</center>
$$11\cdot 52.011\leq s_3\leq 15\cdot 52.011$$
<center>ram</center>
$$11\cdot 8.2869\leq r_3\leq 15\cdot 8.2869$$
<center>hd</center>
$$11\cdot 416.6017\leq h_3\leq 15\cdot 416.6017$$

---

<center>price</center>
$$4\cdot 1492.1775\leq p_{4}\leq 13\cdot 1492.1775$$
<center>speed</center>
$$4\cdot 52.011\leq s_4\leq 13\cdot 52.011$$
<center>ram</center>
$$4\cdot 8.2869\leq r_4\leq 13\cdot 8.2869$$
<center>hd</center>
$$4\cdot 416.6017\leq h_4\leq 13\cdot 416.6017$$


---

$$p_{1} + 5.8112\cdot s_{1} + 78.2357\cdot r_{1} - 0.5357\cdot h_{1} \leq 10000$$
$$s_{2} + r_{2} \geq 300$$
$${𝑟𝑎𝑚}_3 \leq 8$$
$${hd}_4 \geq 1000$$
$$s_{4} + r_{4} \geq 700$$

# Оптимизация методом последовательных уступок

In [114]:
install.packages("lpSolve")

In [115]:
library(lpSolve)

In [116]:
# целевая функция
objective.in <- c(1,       5.8112,      78.2357,      -0.5357,
                  1,       5.8112,      78.2357,      -0.5357,
                  1,       5.8112,      78.2357,      -0.5357,
                  1,       5.8112,      78.2357,      -0.5357)

# матрица ограничений
mat<-c(
      # h_2 -> max
      0, 0, 0, 0,
      0, 0.5 * 0.5 * 1/2200, 0.5* 0.5 * 1/ 704, 0,
      0, 0, 0, 0,
      0, 0.5 * 0.6 * 1/1300, 0.5 * 0.4 * 1/416, 0)
for (i in 1:16){
  tmp <- c()
  for (j in 1:16){
    if (j == i){
      tmp <- append(tmp, 1)
    }
    else{
      tmp <- append(tmp, 0)
    }
  }
  mat <- append(mat, tmp)
  mat <- append(mat, tmp)
}

tmp <- c(
      # производительность C_2 >= 300
      0, 0, 0, 0,
      0, 1, 1, 0,
      0, 0, 0, 0,
      0, 0, 0, 0,
      # производительность C_4 >= 800
      0, 0, 0, 0,
      0, 0, 0, 0,
      0, 0, 0, 0,
      0, 1, 1, 0,
      # стоимость <= 15000
      1, 5.8112, 78.2357, -0.5357,
      0, 0, 0, 0,
      0, 0, 0, 0,
      0, 0, 0, 0
)
mat <- append(mat, tmp)
const.mat<-matrix(mat, nrow=36, byrow=TRUE)

# ограничения
tmp <- c("<=", ">=")
signs <- c()

for (i in 1: 16){
  signs <- append(signs, tmp)
}
signs <- append(c(">="), signs)
signs <- append(signs, c(">=", ">=", "<="))
const.dir<-signs

const.rhs<-c(
    0,
    10 * 1492.1775, 6 * 1492.1775, 10 * 52.0110, 6 * 52.0110, 10 * 8.2869, 6 * 8.2869, 10 * 416.6017, 6 * 416.6017,
    22 * 1492.1775, 6 * 1492.1775, 22 * 52.0110, 6 * 52.0110, 22 * 8.2869, 6 * 8.2869, 22 * 416.6017, 6 * 416.6017,
    15 * 1492.1775, 11 * 1492.1775, 15 * 52.0110, 11 * 52.0110, 15 * 8.2869, 11 * 8.2869, 15 * 416.6017, 11 * 416.6017,
    13 * 1492.1775, 4 * 1492.1775, 13 * 52.0110, 4 * 52.0110, 13 * 8.2869, 4 * 8.2869, 13 * 416.6017, 4 * 416.6017,
    300,
    700,
    15000
)

# h минимизируется
Res<-lp("min", objective.in, const.mat, const.dir, const.rhs)
Res

Success: the objective function is 55230.21 

In [117]:
solution <- Res$solution
solution

[1]  8953.0650   312.0660    49.7214  4166.0170  8953.0650   312.0660
 [7]    49.7214  9165.2374 16413.9525   572.1210    91.1559  6249.0255
[13]  5968.7100   666.8524    33.1476  5415.8221

$$Значение H_1 и H_2:$$

In [118]:
HH1 <- function(sol){
  return (sum(
    c(1,       5.8112,      78.2357,      -0.5357,
      1,       5.8112,      78.2357,      -0.5357,
      1,       5.8112,      78.2357,      -0.5357,
      1,       5.8112,      78.2357,      -0.5357) * sol
  ))
}

HH2 <- function(sol){
  return (sum(
    c(0, 0, 0, 0,
      0, 0.5 * 0.5 * 1/2200, 0.5* 0.5 * 1/ 704, 0,
      0, 0, 0, 0,
      0, 0.5 * 0.6 * 1/1300, 0.5 * 0.4 * 1/416, 0) * sol
  ))
}

H1 = HH1(Res$solution)
H2 = HH2(Res$solution)

HH1(Res$solution)
"=========="
HH2(Res$solution)

[1] 55230.21

[1] "=========="

[1] 0.2229442

Реезультаты:
$$H_1 = 55230.20716199$$
$$H_2 = 0.222944154152098$$
$$ p_1 = 8953.065 $$
$$ s_1 = 312.066 $$
$$ r_1 = 49.7214 $$
$$ h_1 = 4166.017 $$
$$ p_2 = 8953.065 $$
$$ s_2 = 312.066 $$
$$ r_2 = 49.7214 $$
$$ h_2 = 9165.2374 $$
$$ p_3 = 16413.9525 $$
$$ s_3 = 572.121 $$
$$ r_3 = 91.1559 $$
$$ h_3 = 6249.0255 $$
$$ p_4 = 5968.71 $$
$$ s_4 = 666.8524 $$
$$ r_4 = 33.1476 $$
$$ h_4 = 5415.8221 $$


Найдем количество компьютеров для каждого сегмента

In [119]:
xs = solution[c(1, 5, 9, 13)]
Ns = xs / 1492.1775
Ns

[1]  6  6 11  4

In [120]:
N1 = Ns[1]
N2 = Ns[2]
N3 = Ns[3]
N4 = Ns[4]

Выбор компьютеров сегмента $C_1$:

Ограничения:
 - суммарная стоимость компьютеров сегмента $C_1$ должна быть не более 15000 у.е.

In [121]:
result1 = c()
for(i in 1: 4){
  result1 <- append(result1, solution[i] / N1)
}
result1 <- result1[-1] #speed, ram, hd,
result1

[1]  52.0110   8.2869 694.3362

In [122]:
C1_dataset <- data.frame(data)
tmp <- c()

for (i in 1: length(rownames(data))){
    point = c(data$speed[i], data$ram[i], data$hd[i])
    tmp <- append(tmp, sqrt(sum((point - result1) ^ 2)))
}
C1_dataset$eucl <- tmp

In [123]:
comp1 <- subset(C1_dataset, eucl == min(C1_dataset$eucl))
View(comp1)

...1 price speed hd  ram screen cd  multi premium ads trend eucl    
4490 4490 2604  50    720 16  15     yes no    yes     205 21    26.87319
4533 4533 2940  50    720 16  17     yes no    yes     205 21    26.87319
4581 4581 2693  50    720 16  15     yes yes   yes     205 21    26.87319
4584 4584 2640  50    720 16  15     yes no    yes     205 21    26.87319
4636 4636 2794  50    720 16  15     yes no    yes     205 21    26.87319
4643 4643 2545  50    720 16  15     yes no    yes     205 21    26.87319
4709 4709 2445  50    720 16  15     yes no    yes     162 22    26.87319
4721 4721 2504  50    720 16  15     yes no    yes     162 22    26.87319
4731 4731 2694  50    720 16  15     yes no    yes     162 22    26.87319
4732 4732 2593  50    720 16  15     yes yes   yes     162 22    26.87319
4807 4807 2840  50    720 16  17     yes no    yes     162 22    26.87319
4852 4852 2540  50    720 16  15     yes no    yes     162 22    26.87319
4930 4930 2740  50    720 16  17     yes no    yes     191 23    26.87319
4935 4935 2493  50    720 16  15     yes yes   yes     191 23    26.87319
4963 4963 2594  50    720 16  15     yes no    yes     191 23    26.87319
5007 5007 2404  50    720 16  15     yes no    yes     191 23    26.87319
5020 5020 2345  50    720 16  15     yes no    yes     191 23    26.87319
5023 5023 2440  50    720 16  15     yes no    yes     191 23    26.87319
5060 5060 2423  50    720 16  15     yes yes   yes     182 24    26.87319
5062 5062 2354  50    720 16  15     yes no    yes     182 24    26.87319
5064 5064 2295  50    720 16  15     yes no    yes     182 24    26.87319
5109 5109 2390  50    720 16  15     yes no    yes     182 24    26.87319
5151 5151 2690  50    720 16  17     yes no    yes     182 24    26.87319
5217 5217 2544  50    720 16  15     yes no    yes     182 24    26.87319
5422 5422 2254  50    720 16  15     yes no    yes     152 26    26.87319
5426 5426 2540  50    720 16  17     yes no    yes     152 26    26.87319
5443 5443 2290  50    720 16  15     yes no    yes     152 26    26.87319
5444 5444 2444  50    720 16  15     yes no    yes     152 26    26.87319
5476 5476 2195  50    720 16  15     yes no    yes     152 26    26.87319
5482 5482 2323  50    720 16  15     yes yes   yes     152 26    26.87319

Найдем компьютер с минимальной ценой:

In [124]:
resultComp1 <- comp1[which.min(comp1$price), ]
resultComp1

...1 price speed hd  ram screen cd  multi premium ads trend eucl    
5476 5476 2195  50    720 16  15     yes no    yes     152 26    26.87319

Выбор компьютеров сегмента $C_2:$
 - суммарная производительность $C_2$ не менее 300 (считаем аддитивной характеристикой от количества входящих ПК)

In [125]:
C2_dataset <- data.frame()

# доп условий нет
for (i in 1: length(rownames(data))){
    C2_dataset <-rbind(C2_dataset ,data[i,])
}

In [126]:
tmp <- c()
for (i in 1: length(rownames(C2_dataset))){
  tmp <- append(tmp, 0.5 * C2_dataset$speed[i] / max(C2_dataset$speed) + 0.5 * C2_dataset$ram[i] / max(C2_dataset$ram))
}
C2_dataset$specs <- tmp

In [127]:
tmp <- c()
for (i in 1: length(rownames(C2_dataset))){
  tmp <- append(tmp, abs(C2_dataset$specs[i] - H2))
}
C2_dataset$specs <- tmp

In [128]:
specs2 <- min(C2_dataset$specs)
comp2 <- subset(C2_dataset, specs == specs2)
View(comp2)

...1 price speed hd  ram screen cd  multi premium ads trend specs      
1    18  2325  33    210 4   15     no  no    yes     94  1     0.004555846
2    19  2095  33    250 4   15     no  no    yes     94  1     0.004555846
3    21  1695  33    130 4   14     no  no    yes     94  1     0.004555846
4    27  1975  33    130 4   14     no  no    yes     94  1     0.004555846
5    31  1920  33    170 4   14     yes no    yes     94  1     0.004555846
6    35  1999  33    170 4   14     no  no    yes     94  1     0.004555846
7    47  2195  33    130 4   14     no  no    yes     94  1     0.004555846
8    50  1995  33    250 4   14     no  no    yes     94  1     0.004555846
9    51  2499  33    170 4   14     no  no    yes     94  1     0.004555846
10   52  2395  33    130 4   14     no  no    yes     94  1     0.004555846
11   54  2190  33    210 4   14     no  no    yes     94  1     0.004555846
12   62  1795  33    170 4   15     no  no    yes     94  1     0.004555846
13   64  2220  33    250 4   14     yes no    yes     94  1     0.004555846
14   69  2199  33    212 4   14     no  no    yes     94  1     0.004555846
15   72  2725  33    210 4   17     no  no    yes     94  1     0.004555846
16   73  1695  33    170 4   14     no  no    yes     94  1     0.004555846
17   88  1795  33    130 4   14     no  no    yes     94  1     0.004555846
18   97  1999  33    170 4   14     no  no    yes     95  2     0.004555846
19   98  1920  33    170 4   14     yes no    yes     95  2     0.004555846
20  100  1695  33    170 4   14     no  no    yes     95  2     0.004555846
21  101  1995  33    250 4   14     no  no    yes     95  2     0.004555846
22  112  2095  33    130 4   14     no  no    yes     95  2     0.004555846
23  119  2499  33    170 4   14     no  no    yes     95  2     0.004555846
24  120  2799  33    230 4   14     no  no    yes     95  2     0.004555846
25  121  1795  33    170 4   14     no  no    yes     95  2     0.004555846
26  122  2199  33    210 4   14     no  no    yes     95  2     0.004555846
27  133  2220  33    250 4   14     yes no    yes     95  2     0.004555846
28  138  2395  33    250 4   15     no  no    yes     95  2     0.004555846
29  148  2199  33    212 4   14     no  no    yes     95  2     0.004555846
30  151  2095  33    250 4   15     no  no    yes     95  2     0.004555846
⋮   ⋮    ⋮     ⋮     ⋮   ⋮   ⋮      ⋮   ⋮     ⋮       ⋮   ⋮     ⋮          
922 5193 1528  33    428 4   17     no  no    yes     182 24    0.004555846
923 5201 1329  33    212 4   15     no  no    yes     182 24    0.004555846
924 5204 1538  33    424 4   14     yes yes   yes     182 24    0.004555846
925 5215 1268  33    428 4   15     no  no    yes     182 24    0.004555846
926 5226 1554  33    214 4   14     no  no    yes     182 24    0.004555846
927 5228 1199  33    428 4   14     no  no    yes     182 24    0.004555846
928 5238 1490  33    214 4   15     no  no    yes     158 25    0.004555846
929 5241 1199  33    210 4   14     no  no    yes     158 25    0.004555846
930 5242 1495  33    214 4   14     no  no    yes     158 25    0.004555846
931 5251 1295  33    214 4   14     no  no    yes     158 25    0.004555846
932 5265 1399  33    210 4   14     yes yes   yes     158 25    0.004555846
933 5270 1644  33    214 4   14     no  no    yes     158 25    0.004555846
934 5276 1518  33    428 4   17     no  no    yes     158 25    0.004555846
935 5279 1490  33    214 4   14     no  no    yes     158 25    0.004555846
936 5287 1590  33    214 4   14     no  no    yes     158 25    0.004555846
937 5291 1590  33    214 4   15     no  no    yes     158 25    0.004555846
938 5299 1199  33    340 4   14     no  no    yes     158 25    0.004555846
939 5324 1195  33    214 4   14     no  no    yes     158 25    0.004555846
940 5361 1268  33    428 4   15     no  no    yes     158 25    0.004555846
941 5372 1554  33    214 4   14     no  no    yes     158 25    0.004555846
942 5373 1395  33    214 4   14     no  no    yes   

Т.к. найденные компьютеры одинаковы по хар-кам выберем наиболее популярный

In [129]:
resultComp2 <- comp2[which.min(comp2$trend), ]
resultComp2

...1 price speed hd  ram screen cd multi premium ads trend specs      
1 18   2325  33    210 4   15     no no    yes     94  1     0.004555846

Выбор компьютеров сегмента $C_3:$
 - максимальный объем ОЗУ для сегмента $C_3$ не более 8

In [130]:
result3 = c()
for(i in 9: 12){
  result3 <- append(result3, solution[i] / N3)
}
result3 <- result3[-1] #speed, ram, hd,
result3

[1]  52.0110   8.2869 568.0932

In [131]:
C3_dataset <- data.frame()

for (i in 1: length(rownames(data))){
  if (data[i, 5] <= 8){
    C3_dataset <-rbind(C3_dataset, data[i,])
  }
}

In [132]:
tmp <- c()

for (i in 1: length(rownames(C3_dataset))){
  point = c(C3_dataset$speed[i], C3_dataset$ram[i], C3_dataset$hd[i])
  tmp <- append(tmp, sqrt(sum((point - result3) ^ 2)))
}
C3_dataset$eucl <- tmp
C3_dataset

...1 price speed hd   ram screen cd  multi premium ads trend eucl     
1     1   1499  25     80  4   14     no  no    yes     94  1     488.8588 
2     2   1795  33     85  2   14     no  no    yes     94  1     483.5080 
3     3   1595  25    170  4   15     no  no    yes     94  1     399.0316 
4     4   1849  25    170  8   14     no  no    no      94  1     399.0086 
5     7   1720  25    170  4   14     yes no    yes     94  1     399.0316 
6     8   1995  50     85  2   14     no  no    yes     94  1     483.1383 
7     9   2225  50    210  8   14     no  no    yes     94  1     358.0990 
8    10   2575  50    210  4   15     no  no    yes     94  1     358.1245 
9    11   2195  33    170  8   15     no  no    yes     94  1     398.5470 
10   12   2605  66    210  8   14     no  no    yes     94  1     358.3665 
11   13   2045  50    130  4   14     no  no    yes     94  1     438.1188 
12   14   2295  25    245  8   14     no  no    yes     94  1     324.2205 
13   15   2699  50    212  8   14     no  no    yes     94  1     356.0990 
14   16   2225  50    130  4   14     no  no    yes     94  1     438.1188 
15   17   1595  33     85  2   14     no  no    yes     94  1     483.5080 
16   18   2325  33    210  4   15     no  no    yes     94  1     358.6231 
17   19   2095  33    250  4   15     no  no    yes     94  1     318.6897 
18   20   4395  66    452  8   14     no  no    yes     94  1     116.9334 
19   21   1695  33    130  4   14     no  no    yes     94  1     438.5265 
20   22   2795  66    130  4   14     no  no    yes     94  1     438.3375 
21   24   2875  50    210  4   17     no  no    yes     94  1     358.1245 
22   25   4195  50    452  8   14     no  no    yes     94  1     116.1110 
23   26   1290  33     80  2   14     no  no    yes     94  1     488.5038 
24   27   1975  33    130  4   14     no  no    yes     94  1     438.5265 
25   28   3995  33    452  8   14     no  no    yes     94  1     117.6399 
26   30   3244  66    245  8   14     no  no    yes     94  1     323.3961 
27   31   1920  33    170  4   14     yes no    yes     94  1     398.5700 
28   32   1995  25    130  4   14     no  no    yes     94  1     438.9461 
29   33   2595  33    210  8   17     no  no    yes     94  1     358.5976 
30   34   2475  50    210  4   15     no  no    yes     94  1     358.1245 
⋮    ⋮    ⋮     ⋮     ⋮    ⋮   ⋮      ⋮   ⋮     ⋮       ⋮   ⋮     ⋮        
4921 6167 1949   66    420 8   15     no  no    no      40  33    148.75274
4922 6171 1754   66    850 8   15     yes no    yes     40  33    282.25379
4923 6172 1445   66   1000 4   14     no  no    yes     40  33    432.15452
4924 6173 1490  100    528 4   15     no  no    yes     40  33     62.68005
4925 6174 1944   66    850 8   15     yes no    yes     40  33    282.25379
4926 6176 1395  100    528 4   14     no  no    yes     40  33     62.68005
4927 6178 1295   66    528 4   14     no  no    yes     40  33     42.67946
4928 6180 1545  100   1000 4   14     no  no    yes     40  33    434.58576
4929 6182 1590   66    528 8   15     no  no    yes     39  34     42.46459
4930 6187 1690  100    528 8   15     no  no    yes     39  34     62.53394
4931 6188 1844  100    528 8   14     no  no    yes     39  34     62.53394
4932 6191 1699   66    630 8   15     no  no    no      39  34     63.46828
4933 6196 1495   66    528 8   14     no  no    yes     39  34     42.46459
4934 6197 1840  100   1200 8   14     no  no    yes     39  34    633.72644
4935 6199 1595  100    528 8   14     no  no    yes     39  34     62.53394
4936 6214 1740   66   1200 8   14     no  no    yes     39  34    632.06166
4937 6215 1554   66    528 8   14     no  no    yes     39  34     42.46459
4938 6216 1654  100    528 8   14     no  no    yes     39  34     62.53394
4939 6220 1744   66    528 8   14     no  no    yes     39  34     42.46459
4940 6224 1790  100   1200 8   14     no  no    yes     39  35    633.72644
4941 6225 1590   66    528 8   15     no  no    yes  

In [133]:
comp3 <- subset(C3_dataset, eucl == min(C3_dataset$eucl))
View(comp3)

...1 price speed hd  ram screen cd  multi premium ads trend eucl    
1  4757 1499  50    545 4   15     yes no    yes     162 22    23.57369
2  4909 1499  50    545 4   15     yes no    yes     191 23    23.57369
3  5158 1758  50    545 4   17     yes no    yes     182 24    23.57369
4  5178 1499  50    545 4   15     yes no    yes     182 24    23.57369
5  5310 1748  50    545 4   17     yes no    yes     158 25    23.57369
6  5349 1499  50    545 4   15     yes no    yes     158 25    23.57369
7  5477 1499  50    545 4   15     yes no    yes     152 26    23.57369
8  5489 1748  50    545 4   17     yes no    yes     152 26    23.57369
9  5539 1838  50    545 4   17     yes no    yes     152 26    23.57369
10 5554 1808  50    545 4   17     yes no    yes     163 27    23.57369
11 5681 1724  50    545 4   17     yes no    yes     163 27    23.57369
12 5694 1499  50    545 4   15     yes no    yes     163 27    23.57369
13 5793 1199  50    545 4   14     no  no    yes     132 28    23.57369
14 5794 1724  50    545 4   17     yes yes   yes     132 28    23.57369
15 5801 1499  50    545 4   15     yes no    yes     132 28    23.57369
16 5819 1808  50    545 4   17     yes no    yes     132 28    23.57369
17 5821 1808  50    545 4   17     yes yes   yes     132 28    23.57369
18 5826 1724  50    545 4   17     yes no    yes     132 28    23.57369
19 5827 1499  50    545 4   15     yes yes   yes     132 28    23.57369
20 5852 1704  50    545 4   17     yes yes   yes     129 29    23.57369
21 5868 1808  50    545 4   17     yes yes   yes     129 29    23.57369
22 5872 1808  50    545 4   17     yes no    yes     129 29    23.57369
23 5881 1499  50    545 4   15     yes yes   yes     129 29    23.57369
24 5924 1499  50    545 4   15     yes no    yes     129 29    23.57369
25 5928 1704  50    545 4   17     yes no    yes     129 29    23.57369
26 5958 1199  50    545 4   14     no  no    yes     129 29    23.57369

In [134]:
resultComp3 <- comp3[which.min(comp3$price), ]
resultComp3

...1 price speed hd  ram screen cd multi premium ads trend eucl    
1 5793 1199  50    545 4   14     no no    yes     132 28    23.57369

 Выбор компьютеров сегмента $C_4:$
 - минимальный объем жесткого диска для сегмента для сегмента $C_4$ не менее 1000
 - Суммарная производительность для С4 - не менее 700 (считаем аддитивной характеристикой от количества входящих ПК)

In [135]:
result4 = c()
for(i in 13: 16){
  result4 <- append(result4, solution[i] / N4)
}
result4 <- result4[-1] #speed, ram, hd,
result4

[1]  166.7131    8.2869 1353.9555

In [136]:
C4_dataset <- data.frame()

for (i in 1: length(rownames(data))){
  if (data[i, 4] >= 1000){
    C4_dataset <-rbind(C4_dataset, data[i,])
  }
}
C4_dataset

...1 price speed hd   ram screen cd  multi premium ads trend
1    491 3990   66   1000 16  14     no  no    yes     139  5   
2    616 3990   66   1000 16  14     no  no    yes     176  6   
3    721 3025   33   1370  8  14     no  no    yes     249  7   
4    901 3895   66   1370  8  14     no  no    yes     249  7   
5    982 4295   33   1100 16  14     no  no    yes     298  8   
6    994 4494   33   1100 16  14     yes no    yes     298  8   
7   1043 4594   50   1100 16  14     yes no    yes     298  8   
8   1049 3025   33   1370  8  14     no  no    yes     298  8   
9   1099 4495   66   1100 16  14     no  no    yes     298  8   
10  1102 3895   66   1370  8  14     no  no    yes     298  8   
11  1123 4694   66   1100 16  14     yes no    yes     298  8   
12  1177 4395   50   1100 16  14     no  no    yes     298  8   
13  1507 5399   66   1200 32  17     yes no    yes     283 10   
14  1525 3385   66   1370  8  14     no  no    yes     283 10   
15  1611 3135   33   1370  8  14     no  no    yes     283 10   
16  1689 2875   33   1370  4  15     no  no    yes     283 10   
17  1785 3125   66   1370  4  15     no  no    yes     283 10   
18  1836 3205   66   1370  8  14     no  no    yes     259 11   
19  1992 5399   66   1200 32  17     yes no    yes     259 11   
20  2043 2975   33   1370  8  14     no  no    yes     259 11   
21  2062 3795   33   1000 16  14     yes no    yes     275 12   
22  2097 5399   66   1200 32  17     yes no    yes     275 12   
23  2103 2785   33   1060  8  14     no  no    yes     275 12   
24  2149 3995   66   1000 16  14     yes no    yes     275 12   
25  2205 3895   50   1000 16  14     yes no    yes     275 12   
26  2283 2985   66   1060  8  14     no  no    yes     275 12   
27  2469 4799   66   1000 16  17     yes no    yes     216 13   
28  3147 3299   66   1000 16  15     yes no    yes     307 16   
29  3178 3399  100   1000 16  15     yes no    yes     307 16   
30  3231 3499  100   1000 16  15     yes no    yes     307 16   
⋮   ⋮    ⋮     ⋮     ⋮    ⋮   ⋮      ⋮   ⋮     ⋮       ⋮   ⋮    
460 6205 2723  100   1200 24  15     yes yes   yes     39  34   
461 6206 2995   66   1600 32  15     yes no    yes     39  34   
462 6207 2654  100   1200 24  15     yes no    yes     39  34   
463 6209 2495   66   1200 24  15     yes no    yes     39  34   
464 6210 3244   75   1600 32  15     yes no    yes     39  34   
465 6212 2623   66   1200 24  15     yes yes   yes     39  34   
466 6213 2890  100   1200 24  17     yes no    yes     39  34   
467 6214 1740   66   1200  8  14     no  no    yes     39  34   
468 6218 2744   66   1200 24  15     yes no    yes     39  34   
469 6219 2785   66   1600 24  15     yes no    yes     39  34   
470 6222 2595  100   1200 24  15     yes no    yes     39  35   
471 6223 2844  100   1200 24  15     yes no    yes     39  35   
472 6224 1790  100   1200  8  14     no  no    yes     39  35   
473 6226 3244   75   1600 32  15     yes no    yes     39  35   
474 6227 3123   75   1600 32  15     yes yes   yes     39  35   
475 6228 2554   66   1200 24  15     yes no    yes     39  35   
476 6231 2890  100   1200 24  17     yes no    yes     39  35   
477 6236 3290   75   1600 32  17     yes no    yes     39  35   
478 6237 2790   66   1200 24  17     yes no    yes     39  35   
479 6238 2845  100   1600 24  15     yes no    yes     39  35   
480 6240 3095  100   1600 32  15     yes no    yes     39  35   
481 6243 2723  100   1200 24  15     yes yes   yes     39  35   
482 6244 2745   66   1600 24  15     yes no    yes     39  35   
483 6245 3054   75   1600 32  15     yes no    yes     39  35   
484 6246 2744   66   1200 24  15     yes no    yes     39  35   
485 6249 2623   66   1200 24  15     yes yes   yes     39  35   
486 6251 2495   66   1200 24  15     yes no    yes     39  35   
487 6252 2995   66   1600 32  15     yes no    yes     39  35   
488 6253 1690   66   1200  8  14     no  no    yes     39  35   
489 6257 2654  100   1200 24  15     ye

In [137]:
tmp <- c()
for (i in 1: length(rownames(C4_dataset))){
  tmp <- append(tmp, 0.6 * C4_dataset$speed[i] / max(C4_dataset$speed) + 0.4 * C4_dataset$ram[i] / max(C4_dataset$ram))
}
C4_dataset$specs <- tmp

In [138]:
tmp <- c()
for (i in 1: length(rownames(C4_dataset))){
  tmp <- append(tmp, abs(C4_dataset$specs[i] - H2))
}
C4_dataset$specs <- tmp

In [139]:
specs4 <- min(C4_dataset$specs)
comp4 <- subset(C4_dataset, specs == specs4)
View(comp4)

...1 price speed hd   ram screen cd multi premium ads trend specs     
1 1689 2875  33    1370 4   15     no no    yes     283 10    0.02505585

In [140]:
resultComp4 <- comp4[which.min(comp4$trend), ]
resultComp4

...1 price speed hd   ram screen cd multi premium ads trend specs     
1 1689 2875  33    1370 4   15     no no    yes     283 10    0.02505585

# Вывод:

Кол-во компьютеров:

In [141]:
Ns

[1]  6  6 11  4

Номера компьютеров:

In [142]:
c(resultComp1$...1	, resultComp2$...1	, resultComp3$...1	, resultComp4$...1	)

[1] 5476   18 5793 1689

 Суммарная стоимость:

In [143]:
N1*resultComp1$price + N2*resultComp2$price + N3*resultComp3$price + N4*resultComp4$price

[1] 51809